aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+            owner: _,
+            allow_ungated_transfer: _,
+            transfer_events,
+        } = object;
+        event::destroy_handle(transfer_events);
+    }
+
+    // Extension helpers
+
+    public fun generate_extend_ref(ref: &CreatorRef): ExtendRef {
+        ExtendRef { self: ref.self }
+    }
+
+    /// Create a signer for the ExtendRef
+    public fun generate_signer_for_extending(ref: &ExtendRef): signer {
+        create_signer(ref.self.inner)
+    }
+
+    // Transfer functionality
+
+    public fun generate_transfer_ref(ref: &CreatorRef): TransferRef {
+        TransferRef { self: ref.self }
+    }
+
+    public fun generate_linear_transfer_ref(ref: TransferRef): LinearTransferRef {
+        LinearTransferRef { self: ref.self }
+    }
+
+    public fun transfer_with_ref(ref: LinearTransferRef, to: address) acquires Object {
+        let object = borrow_global_mut<Object>(ref.self.inner);
+        event::emit_event(
+            &mut object.transfer_events,
+            TransferEvent {
+                object_id: ref.self,
+                from: object.owner,
+                to,
+            },
+        );
+        object.owner = to;
+    }
+
+    public entry fun entry_transfer(
+        owner: &signer,
+        object_id: address,
+        to: address,
+    ) acquires Object {
+        transfer(owner, ObjectId { inner: object_id }, to)
+    }
+
+    public fun transfer(
+        owner: &signer,
+        object_id: ObjectId,
+        to: address,
+    ) acquires Object {
+        let owner_address = signer::address_of(owner);
+        verify_ungated_and_descendant(owner_address, object_id.inner);
+
+        let object = borrow_global_mut<Object>(object_id.inner);
+        if (object.owner == to) {
+            return
+        };
+
+        event::emit_event(
+            &mut object.transfer_events,
+            TransferEvent {
+                object_id: object_id,
+                from: object.owner,
+                to,
+            },
+        );
+        object.owner = to;
+    }
+
+    public fun transfer_to_object(
+        owner: &signer,
+        object_id: ObjectId,
+        to: ObjectId,
+    ) acquires Object {
+        transfer(owner, object_id, to.inner)
+    }
+
+    /// This checks that the destination address is eventually owned by the owner and that each
+    /// object between the two allows for ungated transfers.
+    fun verify_ungated_and_descendant(owner: address, destination: address) acquires Object {
+        let current_address = destination;
+        assert!(
+            exists<Object>(current_address),
+            error::not_found(EOBJECT_DOES_NOT_EXIST),
+        );
+
+        while (owner != current_address) {
+            // At this point, the first object exists and so the more likely case is that the
+            // object's owner is not an object. So we return a more sensible error.
+            assert!(
+                exists<Object>(current_address),
```

Comment:
Two times exists when you enter the loop, this can be done more gas efficient.

Reply:
The exists have different errors... we could unroll the first iteration of the loop outside the while statement.